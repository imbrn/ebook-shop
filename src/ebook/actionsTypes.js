export const FETCH_ALL_EBOOKS = "FETCH_ALL_EBOOKS";
