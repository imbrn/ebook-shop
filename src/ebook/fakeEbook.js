import Ebook from "./Ebook";

export function fakeEbook(id) {
  return new Ebook({
    id,
    name: `Ebook ${id}`,
    description: `The ebook ${id}`,
    price: id * 10000
  });
}

export function fakeEbooks(amount = 10, firstId = 1) {
  return Array(amount)
    .fill()
    .map(() => fakeEbook(firstId++));
}
