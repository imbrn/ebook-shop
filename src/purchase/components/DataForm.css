.stepPanelContent {
  display: flex;
  flex-wrap: wrap;
}

.personalDataName {
  width: 100%;
}

.personalDataEmail,
.personalDataCpf {
  flex-grow: 2;
}

.billingAddressZipCode,
.billingAddressState,
.billingAddressCity {
  flex-grow: 3;
  flex-basis: 1;
}

.billingAddressAddress {
  width: 100%;
}

.creditCardCardholderName,
.creditCardCardNumber {
  width: 100%;
}

.creditCardDueDate,
.creditCardCvv {
  flex-grow: 2;
}

.paymentMethodButton {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  flex-grow: 2;
  margin: 10px;
  height: 106px;

  svg {
    width: auto;
    pointer-events: none;
  }
}

.paymentBuyNowButton {
  width: 100%;
  text-transform: uppercase;
  margin: 30px 10px 0;
}
