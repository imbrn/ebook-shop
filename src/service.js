let service;

export function setService(it) {
  service = it;
}

export function getService() {
  return service;
}
