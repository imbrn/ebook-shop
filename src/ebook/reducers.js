import { FETCH_ALL_EBOOKS } from "./actionsTypes";
import { combineReducers } from "redux";

export function ebooks(state = [], action) {
  switch (action.type) {
    case FETCH_ALL_EBOOKS:
      return onFetchAllEbooks(state, action);
    default:
      return state;
  }
}

function onFetchAllEbooks(state, action) {
  switch (action.status) {
    case "success":
      return action.ebooks;
    default:
      return state;
  }
}

export function status(state = { kind: "initial" }, action) {
  switch (action.type) {
    case FETCH_ALL_EBOOKS:
      return onStatus(action);
    default:
      return state;
  }
}

function onStatus(action) {
  switch (action.status) {
    case "fail":
      return {
        kind: "fail",
        cause: action.cause
      };
    default:
      return {
        kind: action.status
      };
  }
}

export default combineReducers({
  items: ebooks,
  status
});
