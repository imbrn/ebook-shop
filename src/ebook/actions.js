import { getService } from "../service";
import { FETCH_ALL_EBOOKS } from "./actionsTypes";

export function fetchAllEbooks() {
  return dispatch => {
    dispatch(onFetchAllEbooks("request"));
    return getService()
      .fetchAllEbooks()
      .then(ebooks => dispatch(onFetchAllEbooks("success", { ebooks })));
  };
}

function onFetchAllEbooks(status, props) {
  return {
    type: FETCH_ALL_EBOOKS,
    status,
    ...props
  };
}
