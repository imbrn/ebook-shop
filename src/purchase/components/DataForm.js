import React, { PureComponent, Fragment } from "react";
import StepPanel, { StepPanelField, Button, Select } from "./StepPanel";
import nanoid from "nanoid";
import { connect } from "react-redux";
import { requestBuy } from "../actions";
import { hasEbooksSelected } from "../selectors";
import { getService } from "../../service";
import personalDataValidation from "../validation/personalData";

import billingAddressValidation, {
  zipCode as zipCodeValidation
} from "../validation/billingAddress";

import creditCardValidation from "../validation/creditCard";

export class DataForm extends PureComponent {
  constructor(props) {
    super(props);

    const values = {
      personalData: {
        name: "",
        email: "",
        cpf: ""
      },
      billingAddress: {
        zipCode: "",
        state: "",
        city: "",
        address: ""
      },
      payment: {
        method: "",
        cardholderName: "",
        cardNumber: "",
        dueDate: "",
        cvv: ""
      },
      ...props.initialValues
    };

    this.state = {
      values,
      touched: {
        personalData: {},
        billingAddress: {},
        payment: {}
      },
      errors: {
        personalData: {},
        billingAddress: {},
        payment: {}
      }
    };

    this.changePaymentMethod = this.changePaymentMethod.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSelectChange = this.handleSelectChange.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    const prevZipCode = prevState.values.billingAddress.zipCode;
    const zipCode = this.state.values.billingAddress.zipCode;

    if (prevZipCode != zipCode && zipCodeValidation.test(zipCode)) {
      this.autoFillAddress(zipCode);
    }
  }

  autoFillAddress(zipCode) {
    getService()
      .searchZipCode(zipCode)
      .then(({ uf, localidade }) => {
        this.setInnerState("values", "billingAddress", {
          state: uf || "",
          city: localidade || ""
        });
      })
      .catch(() => {
        /* Don't do anything */
      });
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        {this.renderPersonalData()}
        {this.renderBillingAddress()}
        {this.renderPayment()}
      </form>
    );
  }

  renderPersonalData() {
    const { values, touched, errors } = this.state;

    const nameId = nanoid(8);
    const emailId = nanoid(8);
    const cpfId = nanoid(8);

    return (
      <StepPanel
        number="1"
        title="Personal data"
        open={this.isPersonalDataOpen()}
      >
        <StepPanelField
          fieldId={nameId}
          name="personalData.name"
          label="Name"
          placeholder="Your name here"
          error={touched.personalData.name && errors.personalData.name}
          value={values.personalData.name}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={emailId}
          name="personalData.email"
          label="E-mail"
          placeholder="Your e-mail here"
          error={touched.personalData.email && errors.personalData.email}
          value={values.personalData.email}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={cpfId}
          name="personalData.cpf"
          label="CPF"
          placeholder="Your CPF here"
          error={touched.personalData.cpf && errors.personalData.cpf}
          value={values.personalData.cpf}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </StepPanel>
    );
  }

  renderBillingAddress() {
    const { values, touched, errors } = this.state;

    const zipCodeId = nanoid(8);
    const stateId = nanoid(8);
    const cityId = nanoid(8);
    const addressId = nanoid(8);

    return (
      <StepPanel
        number="2"
        title="Billing Address"
        open={this.isBillingAddressOpen()}
      >
        <StepPanelField
          fieldId={zipCodeId}
          name={"billingAddress.zipCode"}
          label="Zip code"
          placeholder="14530000"
          error={
            touched.billingAddress.zipCode && errors.billingAddress.zipCode
          }
          value={values.billingAddress.zipCode}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          component={Select}
          options={states}
          fieldId={stateId}
          name={"billingAddress.state"}
          label="State"
          placeholder="Your state here"
          error={touched.billingAddress.state && errors.billingAddress.state}
          value={states.find(it => it.value === values.billingAddress.state)}
          onChange={this.handleSelectChange("billingAddress.state")}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={cityId}
          name={"billingAddress.city"}
          label="City"
          placeholder="Your city here"
          error={touched.billingAddress.city && errors.billingAddress.city}
          value={values.billingAddress.city}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={addressId}
          name={"billingAddress.address"}
          label="Address"
          placeholder="Your full address here"
          error={
            touched.billingAddress.address && errors.billingAddress.address
          }
          value={values.billingAddress.address}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </StepPanel>
    );
  }

  renderPayment() {
    const { values } = this.state;

    return (
      <StepPanel number="3" title="Payment" open={this.isPaymentOpen()}>
        <Button
          type="button"
          data-method="boleto"
          selected={values.payment.method === "boleto"}
          onClick={this.changePaymentMethod}
        >
          Boleto Banc√°rio
        </Button>

        <Button
          type="button"
          data-method="creditCard"
          selected={values.payment.method === "creditCard"}
          onClick={this.changePaymentMethod}
        >
          Credit Card
        </Button>

        {values.payment.method === "creditCard"
          ? this.renderCreditCardPayment()
          : null}

        <Button highlighted type="submit" disabled={!this.isReadyToSend()}>
          Buy now
        </Button>
      </StepPanel>
    );
  }

  renderCreditCardPayment() {
    const { values, touched, errors } = this.state;

    const cardholderNameId = nanoid(8);
    const cardNumberId = nanoid(8);
    const dueDateId = nanoid(8);
    const cvvId = nanoid(8);

    return (
      <Fragment>
        <StepPanelField
          fieldId={cardholderNameId}
          name={"payment.cardholderName"}
          label="Cardholder name"
          placeholder="Your name here"
          error={
            touched.payment.cardholderName && errors.payment.cardholderName
          }
          value={values.payment.cardholderName}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={cardNumberId}
          name={"payment.cardNumber"}
          label="Card number"
          placeholder="1234567891011121"
          error={touched.payment.cardNumber && errors.payment.cardNumber}
          value={values.payment.cardNumber}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={dueDateId}
          name={"payment.dueDate"}
          label="Due date"
          placeholder="12/25"
          error={touched.payment.dueDate && errors.payment.dueDate}
          value={values.payment.dueDate}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />

        <StepPanelField
          fieldId={cvvId}
          name={"payment.cvv"}
          label="CVV"
          placeholder="123"
          error={touched.payment.cvv && errors.payment.cvv}
          value={values.payment.cvv}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </Fragment>
    );
  }

  handleSubmit(e) {
    e.preventDefault();
    this.props.requestBuy(this.state.values);
  }

  handleChange(e) {
    const [section, name] = e.target.name.split(".");
    this.setInnerState("values", section, {
      [name]: e.target.value
    });
  }

  handleSelectChange(selectName) {
    return option => {
      const [section, name] = selectName.split(".");
      this.setInnerState("values", section, {
        [name]: option.value
      });
    };
  }

  handleBlur(e) {
    const [section, name] = e.target.name.split(".");
    this.setInnerState("touched", section, {
      [name]: true
    });
  }

  changePaymentMethod(e) {
    this.setInnerState("values", "payment", {
      method: e.target.dataset.method
    });
  }

  isPersonalDataOpen() {
    return hasEbooksSelected(this.props.purchase);
  }

  isBillingAddressOpen() {
    return (
      this.isPersonalDataOpen() &&
      personalDataValidation.test(this.state.values.personalData)
    );
  }

  isPaymentOpen() {
    return (
      this.isBillingAddressOpen() &&
      billingAddressValidation.test(this.state.values.billingAddress)
    );
  }

  isReadyToSend() {
    const creditCard = { ...this.state.values.payment };
    delete creditCard.method;
    return (
      this.isPaymentOpen() &&
      (this.state.values.payment.method === "boleto" ||
        creditCardValidation.test(creditCard))
    );
  }

  setInnerState(type, section, newValues) {
    this.setState(prevState => {
      const typeValues = prevState[type];
      const sectionValues = typeValues[section];
      return {
        ...prevState,
        [type]: {
          ...typeValues,
          [section]: {
            ...sectionValues,
            ...newValues
          }
        }
      };
    });
  }
}

const states = [
  { label: "Acre", value: "AC" },
  { label: "Alagoas", value: "AL" },
  { label: "Amap√°", value: "AP" },
  { label: "Amazonas", value: "AM" },
  { label: "Bahia", value: "BA" },
  { label: "Cear√°", value: "CE" },
  { label: "Distrito Federal", value: "DF" },
  { label: "Esp√≠rito Santo", value: "ES" },
  { label: "Goi√°s", value: "GO" },
  { label: "Maranh√£o", value: "MA" },
  { label: "Mato Grosso", value: "MT" },
  { label: "Mato Grosso do Sul", value: "MS" },
  { label: "Minas Gerais", value: "MG" },
  { label: "Par√°", value: "PA" },
  { label: "Para√≠ba", value: "PB" },
  { label: "Paran√°", value: "PR" },
  { label: "Pernambuco", value: "PE" },
  { label: "Piau√≠", value: "PI" },
  { label: "Rio de Janeiro", value: "RJ" },
  { label: "Rio Grande do Norte", value: "RN" },
  { label: "Rio Grande do Sul", value: "RS" },
  { label: "Rond√¥nia", value: "RO" },
  { label: "Roraima", value: "RR" },
  { label: "Santa Catarina", value: "SC" },
  { label: "S√£o Paulo", value: "SP" },
  { label: "Sergipe", value: "SE" },
  { label: "Tocantins", value: "TO" }
];

DataForm.defaultProps = {
  requestBuy: () => {}
};

const mapStateToProps = state => ({
  purchase: state.purchase
});

const mapDispatchToProps = dispatch => ({
  requestBuy: data => dispatch(requestBuy(data))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DataForm);
