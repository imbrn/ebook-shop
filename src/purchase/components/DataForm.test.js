import React from "react";
import { render, cleanup, fireEvent, wait } from "react-testing-library";
import { DataForm } from "./DataForm";
import { fakeEbooks } from "../../ebook/fakeEbook";
import { setService } from "../../service";

jest.mock("nanoid", () => {
  let lastId = 0;
  return () => ++lastId;
});

jest.mock("react-select", () => {
  return props => (
    <input
      type="text"
      {...props}
      value={props.value ? props.value.value : ""}
    />
  );
});

afterEach(cleanup);

test("renders correctly", () => {
  const { container } = render(<DataForm />);
  expect(container.firstChild).toMatchSnapshot();
});

test("search address by zip code", async () => {
  const purchase = {
    ebooks: fakeEbooks(2)
  };

  const initialValues = {
    personalData: {
      name: "Customer Name",
      email: "customer@email.com",
      cpf: "12345678910"
    }
  };

  setService({
    searchZipCode: () => {
      return Promise.resolve({
        uf: "SP",
        localidade: "Miguelópolis"
      });
    }
  });

  const { getByLabelText } = render(
    <DataForm purchase={purchase} initialValues={initialValues} />
  );

  fireEvent.change(getByLabelText("Zip code"), {
    target: { value: "14530000" }
  });

  await wait(
    () => {
      expect(getByLabelText("State").value).toBe("SP");
      expect(getByLabelText("City").value).toBe("Miguelópolis");
    },
    { timeout: 10 }
  );
});
