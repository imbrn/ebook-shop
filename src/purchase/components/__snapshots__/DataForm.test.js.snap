// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form>
  <div>
    <header>
      <span
        class=""
      >
        1
      </span>
      <span
        class=""
      >
        Personal data
      </span>
    </header>
  </div>
  <div>
    <header>
      <span
        class=""
      >
        2
      </span>
      <span
        class=""
      >
        Billing Address
      </span>
    </header>
  </div>
  <div>
    <header>
      <span
        class=""
      >
        3
      </span>
      <span
        class=""
      >
        Payment
      </span>
    </header>
  </div>
</form>
`;
